build(ci): ensure correct headers are used on macOS

carlocab opened the pull request:
Currently, the release build picks up headers in
`/Library/Frameworks/Mono.framework/Headers`. You can verify this by downloading the latest nightly build and check the output of `nvim --version`.

These headers might be of a different version from the `libintl` that we link to, so let's avoid usage of them by deleting potential sources of headers that we don't want.

To try to avoid future issues from GitHub adding other frameworks to their CI, we also set `CMAKE_FIND_FRAMEWORK` to `LAST` to make sure that CMake looks for Unix libraries first before looking for macOS-style frameworks.

dundargoc commented:
Hmm, from your description it sounds like CMAKE_FIND_FRAMEWORK should be enough to prevent this issue without removing the headers? Or am I misunderstand it?

carlocab commented:
It _should_ but it doesn't hurt to play it safe, I think. Other than gating the `sudo rm -rf` call behind a check for `GITHUB_ACTIONS` in case someone runs the `build_universal_macos.sh` script locally.

dundargoc commented:
I'm sorry, but I disagree with that assessment. If we can avoid extra code then I think we should go with that option. If we're unsure if a section of code is needed I think we should make sure it is actually needed.

dundargoc commented on .github/workflows/test.yml:
```diff
@@ -143,6 +143,8 @@ jobs:
             flags: -D UNSIGNED_CHAR=ON
           - cc: clang
             runner: macos-12
+            flags: -D CMAKE_FIND_FRAMEWORK=LAST
+            deps_flags: -D CMAKE_FIND_FRAMEWORK=LAST
```
Can this be in the cmakefiles instead? We should try to avoid CI-provider specific solutions, and bake in the solutions to the build system if possible.

carlocab commented:
> I'm sorry, but I disagree with that assessment. If we can avoid extra code then I think we should go with that option. If we're unsure if a section of code is needed I think we should make sure it is actually needed, or wait until it has shown to cause problems before we add it.

Well, I don't use the release builds, so I don't know if it's causing anyone problems. But you're in undefined behaviour territory here: three more lines of code that may or may not be needed to be extra sure you're not running afoul of it is a small price to pay.

In any case, I applied your suggestions. I'm not certain they're sufficient, but they likely are.

carlocab commented:
Actually, now that I think about it, I'm pretty sure that `Mono.framework` partly caused #19127, so if I were you I'd still nuke that directory.

dundargoc commented:
My hesitation with this approach is that it's not really a solution and something we can reuse. If a user happens to have the same problem we can't tell them to nuke half their system.

carlocab commented:
```
  qt.shiboken: (3dinput) Found HOMEBREW_OPT with value: "/opt/homebrew/opt" Assuming homebrew build environment.
  qt.shiboken: (3dinput) Filtering out homebrew include path:  "/opt/homebrew/opt/llvm@15/include"
  qt.shiboken: (3dinput) Filtering out homebrew include path:  "/opt/homebrew/opt/openssl@1.1/include"
  qt.shiboken: (3dinput) Filtering out homebrew include path:  "/opt/homebrew/opt/readline/include"
  qt.shiboken: (3dinput) Filtering out homebrew include path:  "/opt/homebrew/opt/sqlite/include"
  qt.shiboken: (3dinput) Filtering out homebrew include path:  "/opt/homebrew/opt/icu4c/include"
```
https://github.com/Homebrew/homebrew-core/actions/runs/4648214077/jobs/8225808252?pr=127674#step:7:2371

Not saying this is the cause, but just interesting.

dundargoc commented on cmake.deps/CMakeLists.txt:
```diff
@@ -12,12 +12,15 @@ list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake" "${PROJECT_SOURCE_DI
   -D CMAKE_GENERATOR_PLATFORM=${CMAKE_GENERATOR_PLATFORM}
-  -D CMAKE_POSITION_INDEPENDENT_CODE=ON)
+  -D CMAKE_POSITION_INDEPENDENT_CODE=ON
+  -D CMAKE_FIND_FRAMEWORK=${CMAKE_FIND_FRAMEWORK})
```
I'm wondering if there's a problem with always setting this to `LAST` aka skipping the cache variable? I don't have mac experience so please correct me.

carlocab replied:
You could try, but that has the potential to surprise some users. I guess we could set it and see if someone complains.

dundargoc replied:
Hmm, I understand. Should we revert this part of your changes and only set this in the CI? Sorry for the back and forth, trying to come up with a solution that is both maintainable and useful.

carlocab replied:
Sure, no problem. It's all small changes anyway.

carlocab replied:
Just realised we need to keep this to make sure our setting of `CMAKE_FIND_FRAMEWORK` propagates to any dependent builds.

dundargoc commented:
I'm not against removing the system headers, but I'd prefer trying the `CMAKE_FIND_FRAMEWORK` approach first to see if that solves it if you don't mind?

carlocab commented:
> Not saying this is the cause, but just interesting.

Oops, I meant to post this in a different PR. Apologies for the spam.

carlocab commented:
> My hesitation with this approach is that it's not really something we can reuse. If a user happens to have the same problem we can't tell them to nuke half their system.

Welcome to building software on macOS. Nuking half your system (or, well, moving things around so you may as well already) is occasionally given as advice by `brew`. MacPorts sometimes tells you to do this as well in their docs.

But, for reusable advice, short of nuking half your system: you could try `CMAKE_FIND_FRAMEWORK=NEVER`, but that's not something I'd set as a default in `CMakeLists.txt`.

carlocab commented:
> Actually, now that I think about it, I'm pretty sure that `Mono.framework` partly caused #19127, so if I were you I'd still nuke that directory.

Oh, and I just came across this tag I made from the last time I worked on the release build: https://github.com/carlocab/neovim/commit/21d01e94c32703ac80adf1dab98e8217d842a0f6 I had completely forgotten about this.

I ended up not doing that because I _thought_ doing something else fixed it, but apparently not.

carlocab committed:
build(ci): ensure correct headers are used on macOS

Currentyl, the release build picks up headers in
`/Library/Frameworks/Mono.framework/Headers`. You can verify this by
downloading the latest nightly build and checking the output of
`nvim --version`.

These headers are likely to be from a different version of `libintl`
than the one we link to. Let's avoid usage of them by setting
`CMAKE_FIND_FRAMEWORK` to `NEVER`.

dundargoc commented:
Thank you!

carlocab commented:
No problem. But try to give me less of a hard time next time 😛
